backend/amd64/selection.ml

```diff
@@ -164,12 +171,75 @@ let is_immediate n = n <= 0x7FFF_FFFF && n >= -0x8000_0000
 
 let is_immediate_natint n = n <= 0x7FFF_FFFFn && n >= -0x8000_0000n
 
+(* Transforms [if] to allow it to be replaced by [cmovcc] *)
+let if_for_cmov exp =
+  match exp with
+  | Cifthenelse(econd, dbg_c, ((Cvar _) as l), dbg_l, ((Cconst_int _) as r), dbg_r) ->
+    force_bind "right" r
+      (fun right -> Cifthenelse(econd, dbg_c, l, dbg_l, right, dbg_r))
+  | Cifthenelse(econd, dbg_c, ((Cconst_int _) as l), dbg_l, ((Cvar _) as r), dbg_r) ->
+    force_bind "left" l
+      (fun left -> Cifthenelse(econd, dbg_c, left, dbg_l, r, dbg_r)) 
+  | Cifthenelse(econd, dbg_c, ((Cconst_int _) as l), dbg_l, ((Cconst_int _) as r), dbg_r) ->
+    force_bind "left" l
+      (fun left ->
+         force_bind "right" r
+           (fun right -> Cifthenelse(econd, dbg_c, left, dbg_l, right, dbg_r))) 
+  | Cifthenelse(econd, dbg_c,
+                Cop(Cload(mc, mut),[Cvar(v)], dbg_load_l), dbg_l,
+                Cop(Cload(mc', mut'),[Cop(Cadda, [Cvar(v'); Cconst_int(i', dbg_i)], dbg_add_r)], dbg_load_r), dbg_r)
+    when mc = mc' && v = v' && mut = mut' ->
+    let ite = Cifthenelse(econd, dbg_c, Cconst_int(0,Debuginfo.none), dbg_l, Cconst_int(i', dbg_i), dbg_r) in
+    Cop(Cload(mc, mut),[Cop(Cadda, [Cvar(v); ite], dbg_add_r)], dbg_load_r)
```

Comment:
This might be wrong: it appears to be able to commute a `Mutable` load with a possibly side-effecting condition (`econd`)?

Reply:
Good spot! I pushed some changes that should help (and make it more clear what's happening)